### PR TITLE
revamp server class, add tests, add typechecking ci

### PR DESCRIPTION
**Pull request by timmyichen**
This PR does a few things, unfortunately all in one commit because I am a bad person:
* Adds `.env.sample`
* Adds typechecking to the github actions CI flow
* Renames test-related router stuff to sample-related router stuff
* Revamps the Server class
  * Mostly rearranging things to make the it extendable - the TestServer class can now extend from it
* Adds jest testing for the API, plus sample test

```
npm run test-server
```
image

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -26,18 +26,14 @@
...
+      "client/*": ["client/*"]
+    }
   },
   "include": [
-    "server/*",
-    "client/*",
-    "pages/*",
+    "server/**/*",
+    "client/**/*",
+    "pages/**/*"
```
Comment (timmyichen) [inline tsconfig.json:36]: `/*` only goes one level deep, `/**/*` means any level deep

File: server/lib/log.ts [hunk 2]
```diff
@@ -1,4 +1,8 @@
 // Logging system that includes timestamps w/ logs
-export default function log(...args) {
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export default function log(...args: any[]) {
+  if (process.env.NODE_ENV === 'test') {
+    return;
+  }
```
Comment (timmyichen) [inline server/lib/log.ts:6]: This is an example of why having this custom log function is useful - during testing, we can silence the logs so the tests don't get too noisy

File: server/lib/db.ts [hunk 3]
```diff
@@ -9,17 +9,22 @@ interface Users extends Document {
 }
 
 // Database object modeling mongoDB data
-class Database {
+export class Database {
+  dbName: string;
   client = new MongoClient(process.env.DB_STRING);
 
+  constructor(dbName = 'virtual-office') {
```
Comment (timmyichen) [inline server/lib/db.ts:16]: Added support for setting the database name, so during testing we can specify a separate database

File: jest.config.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  ],
+  transform: {
+    '^.+\\.(ts|tsx)$': 'ts-jest',
+  },
+  // this is deprecated but the suggested fix doesn't work :|
+  globals: {
+    'ts-jest': {
+      diagnostics: false,
+    },
+  },
```
Comment (timmyichen) [inline jest.config.js:18]: this block is here to _not_ fail tests if there are TS errors - when running tests, we get the message:
```
ts-jest[ts-jest-transformer] (WARN) Define `ts-jest` config under `globals` is deprecated. Please do
transform: {
    <transform_regex>: ['ts-jest', { /* ts-jest config goes here in Jest */ }],
},
```

but adding the block to the `transform` field above doesn't work (still errors on TS errors)

**Review by MatthewBozin — APPROVED** [review]
This looks great! I really appreciate the comments!

**Pull request merged**